### PR TITLE
RFC: use github oauth2 for user login and API authentication

### PR DESCRIPTION
**Pull request by timfeirg**
用 authlib 重写了登录这块, 和之前一样的流程, 只不过是换成 github.

authlib 的文档我不确定完全看懂了, 也不知道是不是这么写的..看文档感觉就是这个意思了.. 

在未登录情况下, 访问任何一个需要登录的 citadel API, 会直接返回302到 github 进行登录和授权, 所以 citadel-web 直接 relay 这个302请求就好 @tonicbupt

### CODE REVIEW

**Comment by timfeirg** [comment]
I guess it wouldn't hurt though, it seems I was just copying code from your docs there, thanks to your great library, I can do this without any actual knowledge of the github oauth API.

And thank you for examining my code. @lepture

File: citadel/config.py [hunk 1]
```diff
@@ -43,13 +42,11 @@
...
-OAUTH2_BASE_URL = getenv('OAUTH2_BASE_URL', default='https://sso.ricebook.net/oauth/api/')
-OAUTH2_ACCESS_TOKEN_URL = getenv('OAUTH2_ACCESS_TOKEN_URL', default='https://sso.ricebook.net/oauth/token')
-OAUTH2_AUTHORIZE_URL = getenv('OAUTH2_AUTHORIZE_URL', default='https://sso.ricebook.net/oauth/authorize')
-OAUTH2_CLIENT_ID = getenv('OAUTH2_CLIENT_ID', default='whatever')
-OAUTH2_CLIENT_SECRET = getenv('OAUTH2_CLIENT_SECRET', default='whatever')
-AUTH_AUTHORIZE_URL = getenv('AUTH_AUTHORIZE_URL', default='https://sso.ricebook.net/auth/profile')
-AUTH_GET_USER_URL = getenv('AUTH_GET_USER_URL', default='https://sso.ricebook.net/auth/user')
+OAUTH_APP_NAME = 'github'
+GITHUB_CLIENT_KEY = 'ce5e0d16937ca68c3f53'
+GITHUB_CLIENT_SECRET = 'e3f80ef7abf3446e63063bfdb211414824c923fe'
```
Comment (lepture) [inline citadel/config.py:47]: Expose secrets?

File: citadel/api/user.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@bp.route('/login')
+def login():
+    if fetch_token(OAUTH_APP_NAME):
+        return jsonify(g.user.to_dict())
+    url, state = oauth.github.generate_authorize_redirect(
+        url_for('user.authorized', _external=True)
+    )
+    session['state'] = state
```
Comment (lepture) [inline citadel/api/user.py:30]: No need for this, Authlib will handle it automatically.

File: citadel/config.py (hunk 1 above)
Comment (timfeirg) [reply citadel/config.py:47]: yeeeeeeeeeeeeeeeeah I created this test app with `scope:email` for like, just for the purpose of demonstration, didn't think there's much to exploit there. (but, is there?)

File: citadel/api/user.py (hunk 2 above)
Comment (timfeirg) [reply citadel/api/user.py:30]: Oh cool, it really does.
BTW I took this line from [this example](https://docs.authlib.org/en/latest/client/mixed.html?highlight=session%5B%27state%27%5D#id1), I mean, if you're surveying how folks are using Authlib or anything like that.

File: citadel/config.py (hunk 1 above)
Comment (lepture) [reply citadel/config.py:47]: @timfeirg yeah, hackers can use your key and secret to do bad things. GitHub will think the bad guy is you.

File: citadel/api/user.py (hunk 2 above)
Comment (lepture) [reply citadel/api/user.py:30]: @timfeirg you are supposed to read https://docs.authlib.org/en/latest/client/frameworks.html#module-authlib.flask.client

Comment (lepture) [reply citadel/api/user.py:30]: There is a note:

```
This OAuthClient is designed for framework integrations, you won’t use it in daily life.
```

Comment (lepture) [reply citadel/api/user.py:30]: I've updated documentation on mixed OAuthClient:

> This `OAuthClient` is designed for developing framework integrations. If you are using Flask or Django, you should read [Integrated Frameworks](https://docs.authlib.org/en/latest/client/frameworks.html#client-frameworks).

https://docs.authlib.org/en/latest/client/mixed.html

Comment (timfeirg) [reply citadel/api/user.py:30]: 真是一位才华横溢的开发者, 向你表示致敬.

**Pull request merged**